Sources/Sentry/SentrySerialization.m

```diff
@@ -0,0 +1,166 @@
+            }
+            [serializedData setValue:[NSNumber numberWithUnsignedInteger:envelope.items[i].header.length] forKey:@"length"];
+        }
+        NSData *itemHeader = [SentrySerialization dataWithJSONObject:serializedData options:opt error:error];
+        if (nil == itemHeader) {
+            [SentryLog logWithMessage:[NSString stringWithFormat:@"Envelope item header cannot be converted to JSON."] andLevel:kSentryLogLevelError];
+            if (error) {
+                *error = NSErrorFromSentryError(kSentryErrorJsonConversionError, @"Envelope item header cannot be converted to JSON");
+            }
+            return nil;
+        }
+        [envelopeData appendData:itemHeader];
+        [envelopeData appendData:[@"\n" dataUsingEncoding:NSUTF8StringEncoding]];
+        [envelopeData appendData:envelope.items[i].data];
+    }
+
+    return envelopeData;
+}
+
++ (SentryEnvelope *_Nullable) envelopeWithData:(NSData *)data {
+    SentryEnvelopeHeader *envelopeHeader = nil;
+    const unsigned char *bytes = [data bytes];
+    int envelopeHeaderIndex = 0;
+    for (int i = 0; i < data.length; ++i) {
+        if (bytes[i] == '\n') {
+            envelopeHeaderIndex = i;
+            // Envelope header end
+            NSData *headerData = [NSData dataWithBytes:bytes length:i];
+#ifdef DEBUG
+            NSString *headerString = [[NSString alloc] initWithData:headerData encoding:NSUTF8StringEncoding];
+            [SentryLog logWithMessage:[NSString stringWithFormat:@"Header %@", headerString] andLevel:kSentryLogLevelDebug];
+#endif
+            NSError *error = nil;
+            NSDictionary *headerDictionary = [NSJSONSerialization JSONObjectWithData:headerData options:0 error:&error];
+            if (nil != error) {
+                [SentryLog logWithMessage:[NSString stringWithFormat:@"Failed to parse envelope header %@", error] andLevel:kSentryLogLevelError];
+            } else {
+                NSString *_Nullable eventId = [headerDictionary valueForKey:@"eventId"];
+                envelopeHeader = [[SentryEnvelopeHeader alloc] initWithId:eventId];
+            }
+            break;
+        }
+    }
+    if (nil == envelopeHeader) {
+        [SentryLog logWithMessage:[NSString stringWithFormat:@"Invalid envelope. No header found."] andLevel:kSentryLogLevelError];
+        return nil;
+    }
+    NSAssert(envelopeHeaderIndex > 0, @"EnvelopeHeader was parsed, its index is expected.");
+    // Parse items
+    NSInteger itemHeaderStart = envelopeHeaderIndex + 1;
+
+    NSMutableArray<SentryEnvelopeItem *> *items = [NSMutableArray new];
+    int endOfEnvelope = data.length - envelopeHeaderIndex;
+    for (NSInteger i = itemHeaderStart; i <= endOfEnvelope; ++i) {
+        if (bytes[i] == '\n' || i == endOfEnvelope) {
+            if (endOfEnvelope == i) i++; // 0 byte attachment
+            NSData *itemHeaderData = [data subdataWithRange:NSMakeRange(itemHeaderStart, i - itemHeaderStart)];
+#ifdef DEBUG
```

Comment:
should we not look for `options.debug` instead?

Reply:
Not sure we need to print the whole envelope in prod builds if Options.Debug = true.
This was just while building the SDK itself in Debug (while writing the parsing code).

Reply:
yep, agreed, if its printing plain text, it won't make sense